xla/service/gpu/gpu_windowed_einsum_handler.cc

```diff
@@ -1142,6 +1143,44 @@ absl::StatusOr<bool> GpuWindowedEinsumHandler::Run(
     changed |= visitor.changed();
   }
 
+  if (all_windowed_einsum_loops.size() > 0) {
+    TF_ASSIGN_OR_RETURN(bool applied_algsimp,
+                        AlgebraicSimplifier(AlgebraicSimplifierOptions())
+                            .Run(module, execution_threads));
+    changed |= applied_algsimp;
+    TF_ASSIGN_OR_RETURN(bool applied_cf,
+                        HloConstantFolding().Run(module, execution_threads));
+    changed |= applied_cf;
+  }
+  for (HloInstruction* loop : all_windowed_einsum_loops) {
+    VLOG(5) << "Processing " << loop->ToString() << " for unrolling.";
+    std::string original_body_name = std::string(loop->while_body()->name());
+    std::string original_cond_name =
+        std::string(loop->while_condition()->name());
+
+    TF_ASSIGN_OR_RETURN(
+        UnrollResult result,
+        WhileLoopUnroller::UnrollAndReturnReplacement(
+            loop, /*unroll_factor=*/-1, /*wrap_in_trivial_loop=*/true,
+            /*force_unroll=*/false));
+
+    if (result.unrolled) {
+      result.new_while_op->while_body()->SetAndSanitizeName(
+          absl::StrCat("unrolled_", original_body_name));
+      result.new_while_op->while_condition()->SetAndSanitizeName(
+          absl::StrCat("unrolled_", original_cond_name));
+      // The loop is fully unrolled but has a trip count of 1
+      // To prevent it from being inlined by while loop simplifier,
```

Comment:
whats the rationale for preserving while loops with trip count of 1?

Reply:
This is for containing the CM loop in a separate computation to get perfect overlap of computes and comms. If the loop is inlined, it might have undesired schedule.

Reply:
Why do we get an undesired schedule if the loop is inlined? Can we investigate this? I don't think we should rely on not inlining a trivial loop.

Reply:
The overlap of gemm+gemm and gemm+comm needs some explicit scheduling of instructions to maximize perf. If we inline this into the parent computation, other passes like collective combiner or fusions passes will fuse the CM instructions with others which introduce undesired data dependencies and schedule.

Reply:
> If we inline this into the parent computation, other passes like collective combiner or fusions passes will fuse the CM instructions with others which introduce undesired data dependencies and schedule.

That means this is a workaround for other passes not doing the right thing, is it? Can we instead fix these other passes?

Reply:
It's more like avoiding the other optimization passes to mangle with CM schedule. A CM loop contains ops like DUS or DS, the fusion passes will naturally fuse those with other ops and create fusion kernels that wont make the gemm-gemm overlap very efficient. We could annotate the CM ops and make all other passes aware of those but it seems like an overkill. This approach is to make CM ops more contained. I think we discussed replacing the trivial loop with a generic call instruction, that refactoring is in progress and will be a follow-up pr to this.